Bottom: 9fd287607b74155a2e66ec1a8f734d11d5999418
Top:    df50358050350d2efc28a8e305fcd1fe8d9cb1b5
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-05-15 10:47:39 +0800

virtio: fix for event index

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/virtio/virtio_ring.c b/drivers/virtio/virtio_ring.c
index d543404..9ebe283 100644
--- a/drivers/virtio/virtio_ring.c
+++ b/drivers/virtio/virtio_ring.c
@@ -982,6 +982,24 @@ static inline int virtqueue_add_packed(struct virtqueue *_vq,
 	return -EIO;
 }
 
+static bool vring_packed_need_event(struct vring_virtqueue *vq,
+				    __u16 off_wrap, __u16 new,
+				    __u16 old)
+{
+    bool wrap = vq->wrap_counter;
+    int off = off_wrap & ~(1 << 15);
+
+    if (new < old) {
+	    new += vq->vring_packed.num;
+	    wrap ^= 1;
+    }
+
+    if (wrap != off_wrap >> 15)
+	    off += vq->vring_packed.num;
+
+    return vring_need_event(off, new, old);
+}
+
 static bool virtqueue_kick_prepare_packed(struct virtqueue *_vq)
 {
 	struct vring_virtqueue *vq = to_vvq(_vq);
@@ -1011,7 +1029,8 @@ static bool virtqueue_kick_prepare_packed(struct virtqueue *_vq)
 #endif
 
 	if (flags == VRING_EVENT_F_DESC)
-		needs_kick = vring_need_event(off_wrap & ~(1<<15), new, old);
+		needs_kick = vring_packed_need_event(vq, off_wrap,
+						     new, old);
 	else
 		needs_kick = (flags != VRING_EVENT_F_DISABLE);
 	END_USE(vq);
